rotkehlchen/exchanges/okx.py

```diff
@@ -0,0 +1,486 @@
+            raise RemoteError(f'{self.name} API request failed due to {str(e)}') from e
+        try:
+            json_response = json.loads(response.text)
+        except JSONDecodeError as e:
+            raise RemoteError(
+                f'{self.name} returned invalid JSON response: {response.text}',
+            ) from e
+
+        return json_response
+
+    def _api_query_list(
+            self,
+            endpoint: Literal['currencies', 'balance', 'trades', 'deposits', 'withdrawals'],
+            options: Optional[dict] = None,
+    ) -> list:
+        """
+        Makes an API query and parses the response into a list
+
+        May raise
+        - RemoteError if the API returns an unexpected response
+        """
+        response = self._api_query(endpoint=endpoint, options=options)
+        data = response.get('data')
+        if data is None or not isinstance(data, list):
+            raise RemoteError(
+                f'{self.name} json response does not contain list `data`: {response}',
+            )
+        return data
+
+    def _api_query_list_paginated(
+            self,
+            endpoint: Literal['currencies', 'balance', 'trades', 'deposits', 'withdrawals'],
+            pagination_key: str,
+            options: Optional[dict] = None,
+    ) -> list:
+        """
+        Makes subsequent API queries until response list length is less than MAX_RESULTS
+        Pagination is handled by setting `pagination_key` value in the `after` query parameter
+        This returns items with `pagination_key` value less than the passed value
+
+        May raise
+        - RemoteError if the API returns an unexpected response
+        """
+        options = options.copy() if options else {}
+        all_items = []
+
+        while True:
+            data = self._api_query_list(endpoint=endpoint, options=options)
+            all_items.extend(data)
+            if len(data) < self.MAX_RESULTS:
+                break
+
+            earliest_item = data[-1]
+            options.update({'after': earliest_item[pagination_key]})
+
+        return all_items
+
+    def validate_api_key(self) -> tuple[bool, str]:
+        response = self._api_query(endpoint='balance')
+        if response.get('code') != '0':
+            return False, 'Error validating API credentials'
+        return True, ''
+
+    def query_balances(self, **kwargs: Any) -> ExchangeQueryBalances:
+        """
+        https://www.okx.com/docs-v5/en/#rest-api-account-get-balance
+
+        May raise
+        - RemoteError if the OKX API or price oracle returns an unexpected response
+        """
+        data = self._api_query_list(endpoint='balance')
+        try:
+            currencies_data = data[0]['details']
+        except (IndexError, KeyError, TypeError) as e:
+            raise RemoteError(
+                f'{self.name} balance API request failed due to unexpected response {str(e)}',
+            ) from e
```

Comment:
1. Why TypeError? We never handle this anywhere. When can this be raised?
2. Just like we do in the rest of the code for KeyError you need better explanation.

```suggestion
        except (IndexError, KeyError, TypeError) as e:
            msg = str(e)
            if isinstance(e, KeyError):
                msg = f'Missing key: {msg}'
            raise RemoteError(
                f'{self.name} balance API request failed due to unexpected response {msg}',
            ) from e
```

Reply:
`TypeError` will be thrown on accessing `data[0]['details']` if `data[0]` is not a dictionary